### ISSUE TITLE
Port imap4-utf-7 codec implementation to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6289]
|[<img alt="berdario's avatar" src="https://avatars.githubusercontent.com/u/752835?s=50" width="50" height="50">](https://github.com/berdario)| @berdario reported|
|-|-|
|Trac ID|trac#6289|
|Type|enhancement|
|Created|2013-01-31 22:05:18Z|

I started by porting the imap4-utf-7 codec to Python3
Then I ported test_imap.py by looking at all the obvious syntax changes

I did the same with the rest of imap4.py, while keeping it to run correctly with Python2
At then end I fixed the IMAP4UTF7TestCase to run with Python3

So, this patch ended up being a little bit bigger than I expected it to be... when working on it I committed some incremental changes on my bazaar branch, if you want to look at it, it's here:
https://code.launchpad.net/~berdario/twisted/imap4-py3

Attachments:

* [imap4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/38e/38eb05e94b3791a75949b3d54ad354e9d8e74f15/c1ae833211cf2601568f81218512337ff1c4ae59.patch) (54593 bytes) - added by berdario on 2013-01-31 22:05:41Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6289 6289
type__enhancement enhancement
reporter__berdario berdario
priority__normal normal
milestone__Python_3_x Python-3.x
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1359669918000000 1359669918000000
changetime__1474264551168303 1474264551168303
version__None None
owner__berdario berdario
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="berdario's avatar" src="https://avatars.githubusercontent.com/u/752835?s=50" width="50" height="50">](https://github.com/berdario)<a name="note_1"></a>|@berdario commented|
|-|-|

3 details I forgot before:

- the patch also contains a new .bzrignore, you probably don't want it (but since it's part of the bazaar branch, it got included)

- the patch has been created with bzr send -o - as instructed here: http://twistedmatrix.com/trac/wiki/BazaarMirror

- my commit 16492 message's "now the tests pass" doesn't really mean anything: since the module wasn't added to _twistedpython3.py it was passing 0 out 0 tests (so yes: technically it was passing all of them :) )

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @berdario|
|-|-|

  1. It's not clear from the description if this is actually a full port of the module to Python 3. Is it? Besides anything you know you haven't ported, if you run the tests under coverage.py with branch coverage, are parts of the code not tested?
  1. imap module should be added to ported modules list (if it's ported), and cred modules to list of unported but necessary modules (since you're just getting them to import).
  1. If you're going to expand the public API of `compat` you need to add `intTypes` to `__all__` and also document it (in this case in the module docstring). And have a test, for that matter. At which point that probably should be a separate ticket.

(I'm tempted to suggest the IMAP code also be split into multiple tickets, but want to hear the answer to the first question.)

**Comment by twisted-trac** [comment]
|[<img alt="berdario's avatar" src="https://avatars.githubusercontent.com/u/752835?s=50" width="50" height="50">](https://github.com/berdario)<a name="note_4"></a>|@berdario commented|
|-|-|

1. No, it's not a full port... I run coverage.py and indeed there're 3 lines in the encoding/decoding functions that aren't tested: 6177, 6178, 6217
2. the imap module has already been added to the ported modules (but it's not entirely ported)... I need to add cred to the almostModules
3. yes, I forgot about compat.__all__ and the other things

**Comment by twisted-trac** [comment]
|[<img alt="berdario's avatar" src="https://avatars.githubusercontent.com/u/752835?s=50" width="50" height="50">](https://github.com/berdario)<a name="note_6"></a>|@berdario removed owner|
|-|-|

After 4 years, I see that now Twisted has some Python3 compatibility (I'm actually using it... but upgrading is a pain, since support is still unstable) and some of the changes I did have been applied already.

This made it simple to recreate the changes to make IMAP4UTF7TestCase pass, without changing too much code

https://github.com/twisted/twisted/pull/541

I tested my changes on both Python3.5 and Python2.7 by running

py.test test_imap.py -k UTF7

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_7"></a>|@rodrigc set owner to @berdario|
|-|-|

https://github.com/twisted/twisted/pull/541#pullrequestreview-479643

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_8"></a>|@rodrigc set status to `closed`|
|-|-|

In changeset 9fd621c0d425661bf23332cd1692ed256aa2e0b9
```
#!CommitTicketReference repository="" revision="9fd621c0d425661bf23332cd1692ed256aa2e0b9"
Merge pull request #541 from berdario/py3-imap4-take2

Author: berdario
Reviewer: rodrigc
Fixes: #6289

Port imap4-utf-7 codec implementation to Python 3
```
